dbms/src/AggregateFunctions/AggregateFunctionTopK.cpp

```diff
@@ -20,17 +20,24 @@ class AggregateFunctionTopKDateTime : public AggregateFunctionTopK<DataTypeDateT
 };
 
 
-static IAggregateFunction * createWithExtraTypes(const IDataType & argument_type)
+static IAggregateFunction * createWithExtraTypes(const DataTypes & argument_types)
 {
+    const IDataType& argument_type = *argument_types[0];
+
          if (typeid_cast<const DataTypeDate *>(&argument_type))     return new AggregateFunctionTopKDate;
     else if (typeid_cast<const DataTypeDateTime *>(&argument_type))    return new AggregateFunctionTopKDateTime;
     else
     {
+        IAggregateFunction* fun;
+
         /// Check that we can use plain version of AggregateFunctionTopKGeneric
         if (argument_type.isValueUnambiguouslyRepresentedInContiguousMemoryRegion())
-            return new AggregateFunctionTopKGeneric<true>;
+            fun = new AggregateFunctionTopKGeneric<true>;
         else
-            return new AggregateFunctionTopKGeneric<false>;
+            fun = new AggregateFunctionTopKGeneric<false>;
+
+        fun->setArguments(argument_types);
```

Comment:
Still don't understand the root cause: why `setArguments` need to be called here and the call in ExpressionAnalyzer is not enough. Need to look further.

Reply:
The moment is called there is too late. Function AggreagteFunctionGroupUniqArrayGeneric::getReturnType is called before that moment and it's crashing because input_data_type It's NULL

Reply:
Stack is here: https://github.com/yandex/ClickHouse/issues/1028

Reply:
It means, that this fix is ok only as temporary solution for few days.
For proper fix, we need to get rid of `setArguments` method and pass arguments in constructor of aggregate function in factory methods.

Reply:
this will be great :) issue #1028 it's so old that even a temporary fix it's good for it :)